selectable_pdf_infra/lambda/selectable_pdf/main.py

```diff
@@ -0,0 +1,180 @@
+from helpertools import (
+    get_logger,
+    load_pdf_from_s3, 
+    save_pdf_to_s3, 
+    make_pdf_doc_searchable
+)
+
+# typing
+from typing import Dict, Optional
+
+
+# logger
+# ------ 
+#If no LOG_LEVEL env var or wrong LOG_LEVEL env var, fallback 
+# to INFO log level
+logger = get_logger(os.getenv('LOG_LEVEL', default='INFO'))
+
+
+# lambda entrypoint
+# -----------------
+def lambda_handler(event, context):
+    logger.info('event: {}'.format(event))
+    args = parse_args(event)
+    logger.info(f'args: {args}')
+
+    # Get AWS connectors.
+    ddb_ress = boto3.resource('dynamodb')
+    ddb_doc_table = ddb_ress.Table(args['ddb_documents_table'])
+
+    returns = list()
+    for rec in args['records']:
+        document_id = rec['document_id']
+        document_name = rec['document_name']
+        logger.info('document id: {}'.format(document_id))
+        logger.info('document name: {}'.format(document_name))
+        logger.info(f"document bucket: {rec['original_document_s3']['bucket']}")
+        logger.info(f"document key: {rec['original_document_s3']['key']}")
+
+        # store info about starting creating the selectable PDF document
+        try:
+            ddb_doc_table.update_item(
+                Key={
+                    'document_id': document_id,  # HASH key
+                    'document_name': document_name  # RANGE key
+                },
+                UpdateExpression='SET selectable_pdf=:att1',  # This will set a new attribute
+                ExpressionAttributeValues={
+                    ':att1': {
+                        'datetime': datetime.datetime.now().strftime('%Y-%m-%dT%H:%M:%S') + '+00:00'
+                    }
+                }
+            )
+        except Exception as ex:
+            logger.error(f'Cannot update item in DynamoDB table {args["ddb_documents_table"]}')
+            raise ex
+
+        # process the PDF
+        pdf_doc = load_pdf_from_s3(rec['original_document_s3']['bucket'], rec['original_document_s3']['key'])
+        textract_blocks = load_json_from_s3(rec['textract_output_s3']['bucket'], rec['textract_output_s3']['key'])
+        textract_blocks = textract_blocks['Blocks']
+        logger.info(f'nb blocks: {len(textract_blocks)}')
+        num_word_blocks = 0
+        for blk in textract_blocks:
+            if blk['BlockType'] == 'WORD':
+                num_word_blocks += 1
+        logger.info(f'number of WORD blocks {num_word_blocks}')
+
+        selectable_pdf_doc = make_pdf_doc_searchable(
+            pdf_doc=pdf_doc,
+            textract_blocks=textract_blocks,
+            add_word_bbox=args['add_word_bbox'],
+            show_selectable_char=args['show_character'],
+            pdf_image_dpi=args['pdf_image_dpi'],
+            verbose=True
+        )
+        output_key = document_name
+        save_pdf_to_s3(selectable_pdf_doc, args['output_bucket'], output_key)
+
+        # store info about ending the selectable PDF document creation
+        try:
```

Comment:
I see this block of code repeating (lines 112-127 and lines 72-86). There is an opportunity to extract this as a function to be reused.

Reply:
done